class Solution 
{
    public int[] searchRange(int[] nums, int target) 
    {
        int first = findFirst(nums, target);
        int last = findLast(nums, target);

        return new int[]{first, last};
    }

    // Find first occurrence
    private int findFirst(int[] nums, int target) 
    {
        int left = 0;
        int right = nums.length - 1;
        int answer = -1;

        while (left <= right) 
        {
            int mid = left + (right - left) / 2;

            if (nums[mid] < target) 
            {
                left = mid + 1;
            } 
            else if (nums[mid] > target) 
            {
                right = mid - 1;
            } 
            else 
            {
                answer = mid;

                // Search further left
                right = mid - 1;
            }
        }

        return answer;
    }

    // Find last occurrence
    private int findLast(int[] nums, int target) 
    {
        int left = 0;
        int right = nums.length - 1;
        int answer = -1;

        while (left <= right) 
        {
            int mid = left + (right - left) / 2;

            if (nums[mid] < target) 
            {
                left = mid + 1;
            } 
            else if (nums[mid] > target) 
            {
                right = mid - 1;
            } 
            else 
            {
                answer = mid;

                // Search further right
                left = mid + 1;
            }
        }

        return answer;
    }
}